src/full_client.toit

```diff
@@ -16,6 +16,49 @@ import .transport
 
 CLIENT_CLOSED_EXCEPTION ::= "CLIENT_CLOSED"
 
+/**
+A class that ensures that the connection to the broker is kept alive.
+
+When necessary sends ping packets to the broker.
+*/
+class ActivityChecker_:
+  connection_ /Connection_
+  keep_alive_ /Duration
+  current_connection_ /Connection_? := null
+
+  constructor .connection_ --keep_alive/Duration:
+    keep_alive_ = keep_alive
+
+  /**
+  Checks for activity.
+
+  Returns a duration for when it wants to be called again.
+  */
+  check -> Duration:
+    // TODO(florian): we should be more clever here:
+    // We should monitor when the transport starts writing, and when it gets a chunk through.
+    // Also, we should monitor that we actually get something from the server.
+    last_write_us := connection_.transport_.last_write_us
+    remaining_keep_alive_us := keep_alive_.in_us - (Time.monotonic_us - last_write_us)
+    if remaining_keep_alive_us > 0:
+      remaining_keep_alive := Duration --us=remaining_keep_alive_us
+      return remaining_keep_alive
+    else if not connection_.is_writing:
+      // TODO(florian): we need to keep track of whether we have sent a ping.
+      connection_.write PingReqPacket
+      return keep_alive_ / 2
+    else:
+      // TODO(florian): we are currently sending.
+      // We should detect timeouts on the sending.
+      return keep_alive_
+
+  run:
+    while not connection_.is_closed:
```

Comment:
Can the task that runs this get canceled? I think you should check for that in the loop condition, even though it probably works as is because you only cancel when you also close.

Reply:
done.